### PR TITLE
new ansible playbook and ec2 user-data script to allow any new instance to provision/deploy itself

### PR DESCRIPTION
**Pull request by joshm1**
@stkem @andrewconner 

I still need to update the README and confluence docs, but this is a new way to launch any instance, including spot instances.  

To get this to work, a new instance must have the kifi-service IAM role. This allows the instance to make API calls to AWS w/o embedding access/secret keys.  https://console.aws.amazon.com/iam/home?region=us-west-1#roles/kifi-service

I also added a new S3 bucket which we can use to host the ansible playbooks, files, etc.

I've launched 2 shoebox spot instances to test this, on top of clean Ubuntu 14.04 AMIs (not using the play AMI), and they are up and running now

image

This was written to be backwards compatible with still being able to launch/update remote servers, but I should consolidate some things because there is some unnecessary duplication.

You can see output from the cloud-init on the servers at /var/log/cloud-init-output.log, which will look like this:

```
TASK: [ricbra.logentries | Follow logs] ***************************************
changed: [localhost] => (item={'path': u'/home/fortytwo/run/shoebox/log/shoebox.out', 'name': u'shoebox.out'})
changed: [localhost] => (item={'path': u'/home/fortytwo/run/shoebox/log/app.log', 'name': 'app.log'})

TASK: [jtyr.ulimit | Create ulimit configuration] *****************************
changed: [localhost]

TASK: [kifi.common | install zsh] *********************************************
changed: [localhost]

TASK: [kifi.common | add users] ***********************************************
changed: [localhost] => (item=eng)
changed: [localhost] => (item=fortytwo)

TASK: [kifi.common | add eng to sudo group] ***********************************
changed: [localhost]

TASK: [kifi.common | create users .ssh directory] *****************************
changed: [localhost] => (item=eng)
changed: [localhost] => (item=fortytwo)

TASK: [kifi.common | set users .ssh/authorized_keys] **************************
changed: [localhost] => (item=eng)
changed: [localhost] => (item=fortytwo)

TASK: [kifi.common | copy .zshrc] *********************************************
changed: [localhost] => (item=eng)
changed: [localhost] => (item=fortytwo)

TASK: [kifi.common | update sudoers, allow nopasswd for "sudo" group] *********
changed: [localhost]

TASK: [kifi.common | ensure ulimit is set by default] *************************
changed: [localhost] => (item=common-session)
changed: [localhost] => (item=common-session-noninteractive)

TASK: [kifi.common | register user forytwo's crontab] *************************
failed: [localhost] => {"changed": false, "cmd": ["crontab", "-u", "fortytwo", "-l"], "delta": "0:00:00.003710", "end": "2015-09-02 18:01:15.584127", "rc": 1, "start": "2015-09-02 18:01:15.580417", "stdout_lines": [], "warnings": []}
stderr: no crontab for fortytwo
...ignoring

TASK: [kifi.common | delete user fortytwo's crontab if ansible not altered it before] ***
skipping: [localhost]

TASK: [kifi.common | install packages] ****************************************
changed: [localhost] => (item=python-software-properties,wget,tmux,tshark,screen,rsync,mosh,htop,gzip,curl,zip,fail2ban,ntp,python-setuptools,python-pip,imagemagick)

TASK: [kifi.common | install nginx package] ***********************************
changed: [localhost]

TASK: [kifi.common | copy nginx.conf] *****************************************
changed: [localhost]

TASK: [kifi.common | copy nginx common.conf] **********************************
changed: [localhost]

TASK: [kifi.common | copy nginx shoebox.conf] *********************************
changed: [localhost]

TASK: [kifi.common | verify correct nginx syntax] *****************************
ok: [localhost]

TASK: [kifi.common | service name=nginx state=reloaded enabled=yes] ***********
changed: [localhost]

TASK: [kifi.common | setup app directories] ***********************************
changed: [localhost] => (item=repo)
changed: [localhost] => (item=run)

TASK: [kifi.common | include variables for shoebox and instance type {{ ec2_instance_type | regex_replace('\W', '_') }}] ***
failed: [localhost] => {"failed": true, "item": null}
msg: No source file given
...ignoring

TASK: [kifi.common | install shoebox init.d script] ***************************
changed: [localhost]

TASK: [kifi.common | stat path=/etc/rc5.d/S20shoebox] *************************
ok: [localhost]

TASK: [kifi.common | update-rc.d shoebox] *************************************
changed: [localhost]

TASK: [kifi.common | enable shoebox service] **********************************
ok: [localhost]

TASK: [kifi.common | Copy eddie deploy script] ********************************
changed: [localhost]

TASK: [kifi.common | set eddie cron] ******************************************
changed: [localhost]

TASK: [kifi.common | check if shoebox has been deployed before] ***************
ok: [localhost]

TASK: [kifi.common | run eddie-self-deploy] ***********************************
changed: [localhost]

TASK: [kifi.common | install log_stats python dependencies] *******************
changed: [localhost] => (item=boto)
changed: [localhost] => (item=iso8601)
changed: [localhost] => (item=portalocker)
ok: [localhost] => (item=python-dateutil)

TASK: [kifi.common | copy log_stats script] ***********************************
changed: [localhost]

TASK: [kifi.common | setup log_stats cron] ************************************
changed: [localhost]

NOTIFIED: [ricbra.logentries | Restart logentries] ****************************
changed: [localhost]

NOTIFIED: [kifi.common | eddie-self-deploy] ***********************************
changed: [localhost]

NOTIFIED: [ricbra.logentries | Restart logentries] ****************************
changed: [localhost]

PLAY RECAP ********************************************************************
localhost                  : ok=58   changed=40   unreachable=0    failed=0
```

### CODE REVIEW

File: provisioning/ec2-user-data [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (joshm1) [inline provisioning/ec2-user-data:1]: this file is critical for self-provision to work; it must be included in the advanced section when creating a new instance or spot request

Comment (stkem) [inline provisioning/ec2-user-data:1]: Could you make a (possibly very) short google doc or wiki page explaining how to do this (maybe with a screenshot) and send it out.

Comment (stkem) [inline provisioning/ec2-user-data:1]: Ah, I'm an idiot. It's in the PR. Just copy/paste that somewhere more persistent :-)

**Comment by stkem** [comment]
:+1:

**Pull request merged**